Jenkinsfile

```diff
@@ -24,36 +48,77 @@ node {
+        if (versionSuffix != "") {
+            newPkgVersion = newPkgVersion + "-" + versionSuffix
         }
+
+        echo "Publishing ${pkgName} v${newPkgVersion} from ${releaseFromBranch} branch."
+
+        nodeEnv.inside("-e HOME=${workspace} -e BRANCH_NAME=${env.BRANCH_NAME}") {
+            withCredentials([
+                string(credentialsId: 'npm-token', variable: 'NPM_TOKEN'),
+                usernamePassword(credentialsId: 'github-jenkins-user',
+                                  passwordVariable: 'GITHUB_TOKEN',
+                                  usernameVariable: 'GITHUB_USERNAME')]) {
+
+                // Configure commit author for version bump commit and auth credentials
+                // for pushing tag to GitHub.
+                //
+                // See https://git-scm.com/docs/git-credential-store
+                sh """
+                git config user.email ${env.GITHUB_USERNAME}@hypothes.is
+                git config user.name ${env.GITHUB_USERNAME}
+                git config credential.helper store
+                echo https://${env.GITHUB_USERNAME}:${env.GITHUB_TOKEN}@github.com >> \$HOME/.git-credentials
+                """
+
+                // Fetch information about tags so that changelog generation script
+                // can produce diff since last tag. Also remove local tags that no
+                // longer exist on the remote.
+                //
+                // The `--prune-tags` option is not supported in Git 2.11 so we
+                // use the workaround from https://github.com/git/git/commit/97716d217c1ea00adfc64e4f6bb85c1236d661ff
+                sh "git fetch --quiet --prune origin 'refs/tags/*:refs/tags/*' "
+
+                // Bump the package version, update the changelog and create the tag
+                // and GitHub release.
+                sh "yarn version --new-version ${newPkgVersion}"
+
+                // Publish the updated package to the npm registry.
+                // Use `npm` rather than `yarn` for publishing.
+                // See https://github.com/yarnpkg/yarn/pull/3391.
+                sh "echo '//registry.npmjs.org/:_authToken=${env.NPM_TOKEN}' >> \$HOME/.npmrc"
+                sh "npm publish --tag ${npmTag}"
+            }
+        }
+
+        echo "Uploading package ${pkgName} v${newPkgVersion} to CDN"
+
+        // Upload the contents of the package to an S3 bucket, which it
+        // will then be served from.
+        docker.image('nickstenning/s3-npm-publish')
+              .withRun('', "${pkgName}@${newPkgVersion} s3://cdn.hypothes.is") { c ->
+                sh "docker logs --follow ${c.id}"
+              }
     }
 }
 
-boolean isTag() {
-    try {
-        sh 'git fetch --tags'
-        sh 'git describe --exact-match --tags'
-        return true
-    } catch (Exception e) {
-        echo e.toString()
-        return false
+// Increment the minor part of a `MAJOR.MINOR.PATCH` semver version.
+String bumpMinorVersion(String version) {
```

Comment:
Is there any reasonable case in which this function would receive a nasty value for `version`? I can see there is validation that it can be split on `.` and has three parts, but can it be trusted that those parts are Integer-a-ble strings? Probably yes, just poking.

Reply:
The input must be a semver version so the second part must be an integer. The third part might not be an integer, eg. if the version is "1.94.3-a-prerelease". Only the second part needs to be numeric.